src/core/server/rendering/rendering_service.tsx

```diff
@@ -103,6 +103,9 @@ export class RenderingService {
       buildNum,
     });
 
+    const publicPlugins = [...uiPlugins.public].filter(
+      ([, plugin]) => includeUserSettings || plugin.enableForAnonymousPages
```

Comment:
I don't think it's fine to only include the plugins that are explicitly `enableForAnonymousPages`, I think we should also include the whole dependency tree for those plugins, both `requiredPlugins` and `requiredBundles` don't we?

Reply:
I'm a bit torn on this.

I think the _safer_ option is to do what we have now, so we don't accidentally introduce more plugin dependencies in the future that wind up in anonymous pages.
If we do introduce another required dependency for the Security plugin in the future (without also marking that plugin as `enabledOnAnonymousPages`), the login page _should_ break, and we can decide if we want to traverse the dependency tree at that point.

WDYT?

For now I'll add a comment here that says we explicitly _don't_ include all required plugins/bundles.

Reply:
( cc @elastic/kibana-core )

Reply:
After a discussion with @jportner, it also appear that most of the security plugin's dependencies are exclusively used on the server-side, however technically the plugin system does not make a distinction between client and server-side dependencies, meaning that all dependencies are implicitly considered used on both side.

If we were to improve the plugin system / manifest to allow defining distinct client or server dependencies, it could also ease the problem a lot.

Reply:
> We may not need to load canvas when the user is in discover

Unfortunately, with the amount of registry patterns used in Kibana, this is very unlikely to be feasible.. As you mentioned, the navLinks are the first example coming to mind, but it's far from being the only one.